ietf/doc/storage_backends.py

```diff
@@ -0,0 +1,137 @@
+# Copyright The IETF Trust 2025, All Rights Reserved
+
+import debug  # pyflakes:ignore
+
+from hashlib import sha384
+from io import BufferedReader
+from storages.backends.s3 import S3Storage
+from storages.utils import is_seekable
+from typing import Dict, Optional, Union
+
+from django.core.files.base import File
+
+from ietf.doc.models import StoredObject
+from ietf.utils.log import log
+from ietf.utils.timezone import timezone
+
+
+class CustomS3Storage(S3Storage):
+
+    def __init__(self, **settings):
+        self.in_flight_custom_metadata: Dict[str, Dict[str, str]] = {}
+        return super().__init__(**settings)
+
+    def store_file(
+        self,
+        kind: str,
+        name: str,
+        file: Union[File, BufferedReader],
+        allow_overwrite: bool = False,
+        doc_name: Optional[str] = None,
+        doc_rev: Optional[str] = None,
+    ):
+        is_new = not self.exists_in_storage(kind, name)
+        # debug.show('f"Asked to store {name} in {kind}: is_new={is_new}, allow_overwrite={allow_overwrite}"')
+        if not allow_overwrite and not is_new:
+            log(f"Failed to save {kind}:{name} - name already exists in store")
+            debug.show('f"Failed to save {kind}:{name} - name already exists in store"')
+            debug.traceback()
+            raise Exception("Not ignoring overwrite attempts while testing")
+        else:
+            try:
+                new_name = self.save(name, file)
+                now = timezone.now()
+                existing_record = StoredObject.objects.filter(store=kind, name=name)
+                if existing_record.exists():
+                    # Note this is updating a queryset which is guaranteed by constraints to have one object
+                    existing_record.update(
+                        sha384=self.in_flight_custom_metadata[name]["sha384"],
+                        len=int(self.in_flight_custom_metadata[name]["len"]),
+                        modified=now,
+                    )
+                else:
+                    StoredObject.objects.create(
+                        store=kind,
+                        name=name,
+                        sha384=self.in_flight_custom_metadata[name]["sha384"],
+                        len=int(self.in_flight_custom_metadata[name]["len"]),
+                        store_created=now,
+                        created=now,
+                        modified=now,
+                        doc_name=doc_name,
+                        doc_rev=doc_rev,
+                    )
+                if new_name != name:
+                    complaint = f"Error encountered saving '{name}' - results stored in '{new_name}' instead."
+                    log(complaint)
+                    debug.show("complaint")
+                    # Note that we are otherwise ignoring this condition - it should become an error later.
```

Comment:
Might as well raise an exception now since we have the log-and-swallow in place

Reply:
leaving this wart for now - need to think through what it should raise when we take away log-and-swallow.